badger/cmd/info.go

```diff
@@ -59,39 +74,139 @@ info. It also prints info about missing/extra files, and general information abo
 files (which are not referenced by the manifest).  Use this tool to report any issues about Badger
 to the Dgraph team.
 `,
-	Run: func(cmd *cobra.Command, args []string) {
-		err := printInfo(sstDir, vlogDir)
-		if err != nil {
-			fmt.Println("Error:", err.Error())
-			os.Exit(1)
+	RunE: handleInfo,
+}
+
+func handleInfo(cmd *cobra.Command, args []string) error {
+	if err := printInfo(sstDir, vlogDir); err != nil {
+		return errors.Wrap(err, "failed to print information in MANIFEST file")
+	}
+
+	// Open DB
+	opts := badger.DefaultOptions
+	opts.TableLoadingMode = options.MemoryMap
+	opts.Dir = sstDir
+	opts.ValueDir = vlogDir
+	opts.ReadOnly = true
+
+	db, err := badger.Open(opts)
+	if err != nil {
+		return errors.Wrap(err, "failed to open database")
+	}
+	defer db.Close()
+
+	if opt.showTables {
+		tableInfo(sstDir, vlogDir, db)
+	}
+
+	if opt.showHistogram {
+		db.PrintHistogram([]byte(opt.withPrefix))
+	}
+
+	if opt.showKeys {
+		if err := showKeys(db, opt.withPrefix); err != nil {
+			return err
 		}
-		if !opt.showTables {
-			return
+	}
+
+	if len(opt.keyLookup) > 0 {
+		if err := lookup(db); err != nil {
+			return errors.Wrapf(err, "failed to perform lookup for the key: %x", opt.keyLookup)
 		}
-		// Open DB
-		opts := badger.DefaultOptions
-		opts.TableLoadingMode = options.MemoryMap
-		opts.Dir = sstDir
-		opts.ValueDir = vlogDir
-		opts.ReadOnly = true
-
-		db, err := badger.Open(opts)
-		if err != nil {
-			fmt.Println("Error:", err.Error())
-			os.Exit(1)
+	}
+	return nil
+}
+
+func showKeys(db *badger.DB, prefix string) error {
+	if len(prefix) > 0 {
+		fmt.Printf("Only choosing keys with prefix: \n%s", hex.Dump([]byte(prefix)))
+	}
+	txn := db.NewTransaction(false)
+	defer txn.Discard()
+
+	iopt := badger.DefaultIteratorOptions
+	iopt.PrefetchValues = false
```

Comment:
Any reason to have this as `false`? We can prefetch items.

Reply:
We don't want to prefetch values. We want to show only the keys.
If we wanted to show values also, then `prefetchValues = true` would've made sense. :)

Reply:
makes sense.